src/borkdude/deps.clj

```diff
@@ -225,43 +225,172 @@ For more info, see:
+               "        if (conn instanceof HttpURLConnection)"
+               "           {((HttpURLConnection) conn).setInstanceFollowRedirects(true);}"
+               "        ReadableByteChannel readableByteChannel = Channels.newChannel(conn.getInputStream());"
+               "        FileOutputStream fileOutputStream = new FileOutputStream(args[1]);"
+               "        FileChannel fileChannel = fileOutputStream.getChannel();"
+               "        fileOutputStream.getChannel().transferFrom(readableByteChannel, 0, Long.MAX_VALUE);"
+               "        System.exit(0);"
+               "    } catch (IOException e) {"
+               "        e.printStackTrace();"
+               "        System.exit(1); }}}"))
+    dest-file))
+
+
+(defn- clojure-tools-download-java
+  "Downloads URL file to DEST-ZIP-FILE by invoking `java` with JVM-OPTS
+  on a `.java` program file, and returns true on success. Requires
+  Java 11+ (JEP 330)."
+  [url dest-zip-file jvm-opts]
+  (let [dest-dir (.getCanonicalPath (io/file dest-zip-file ".."))
+        dlr-path (clojure-tools-java-downloader-spit dest-dir)
+        java-cmd (get-java-cmd)
+        success?* (atom true)]
+    (binding [*exit-fn* (fn
+                          ([exit-code] (when-not (= exit-code 0) (reset! success?* false)))
+                          ([exit-code msg] (when-not (= exit-code 0)
+                                             (warn msg)
+                                             (reset! success?* false))))]
+      (shell-command (vec (concat [java-cmd]
+                                  jvm-opts
+                                  [dlr-path url (str dest-zip-file)])))
+      (io/delete-file dlr-path true)
+      @success?*)))
 
 (defn clojure-tools-jar-download
-  "Downloads clojure tools jar into deps-clj-config-dir."
-  [deps-clj-config-dir]
-  (let [dir (io/file deps-clj-config-dir)
-        zip (io/file deps-clj-config-dir "tools.zip")]
-    (.mkdirs dir)
-    (download (format "https://download.clojure.org/install/clojure-tools-%s.zip" @version)
-              zip)
-    (unzip zip (.getPath dir))
-    (.delete zip))
+  "Downloads clojure tools archive in OUT-DIR, if not already there,
+  and extracts in-place the clojure tools jar file and other important
+  files.
+
+  The download is attempted by the following two methods in order:
+
+  1. via java, invoking a java subprocess with
+  JAVA-ARGS-WITH-CLJ-JVM-OPTS options, or
+
+  2. directly from this process.
+
+  It calls `*exit-fn*` if it cannot download the archive, with
+  instructions how to manually download it."
+  [out-dir java-args-with-clj-jvm-opts]
+  (let [{:keys [ct-error-exit-code ct-url-str ct-zip-name]} @clojure-tools-info*
+        dir (io/file out-dir)
+        zip-file (io/file out-dir ct-zip-name)]
+    (when-not (.exists zip-file)
+      (warn "Downloading" ct-url-str "to" (str zip-file))
+      (.mkdirs dir)
+      (or (do (warn "\nAttempting download using " ::java-subprocess-method...
```

Comment:
I think we can skip the new method when `CLJ_JVM_OPTS` hasn't been set

Reply:
Should we probe the `java` version before trying the java method, like you're doing in the tests?

Reply:
The new method is now only called when `CLJ_JVM_OPTS` is set.

Unfortunately is not straightforward to check the java version number in the absence of `babashka.process`. This is because at least on Java 8 there is only a single `--version` cli option and this sends the output to `stderr`.  Thus `deps/shell-command` cannot be used to retrieve the version unless it is modified. 

Options:
1. Modify `shell-command` to accept an extra `:to-err-string` option to capture the `stderr` output of the subprocess. I can't see an easy way to differentiate between `stdout` (the current return value) and `stderr`, unless it returns a map?
2. Create a new fn to call on `ProcessBulder` to just pick up and return the java version from the err stream.

Reply:
OK, I think we're going to let this slide, we will just demand Java 11 for this to work with CLJ_JVM_OPTS. Thank you.